src/liballoc/collections/btree/set.rs

```diff
@@ -1055,6 +1086,46 @@ impl<'a, T> IntoIterator for &'a BTreeSet<T> {
     }
 }
 
+/// An iterator produced by calling `drain_filter` on BTreeSet.
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+pub struct DrainFilter<'a, T>(
+    btree_map::DrainFilter<'a, T, (), Box<dyn 'a + FnMut(&T, &mut ()) -> bool>>,
+)
+where
+    T: 'a + Ord;
+
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+impl<'a, T> Drop for DrainFilter<'a, T>
+where
+    T: 'a + Ord,
+{
+    fn drop(&mut self) {
+        self.for_each(drop);
+    }
+}
+
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+impl<'a, T> fmt::Debug for DrainFilter<'a, T>
+where
+    T: 'a + Ord + fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("DrainFilter").field(&0).finish()
+    }
+}
+
+#[unstable(feature = "btree_drain_filter", issue = "42849")]
+impl<'a, T> Iterator for DrainFilter<'a, T>
+where
+    T: 'a + Ord,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<T> {
+        self.0.next().map(|(k, _)| k)
```

Comment:
Rather than building on the `map::DrainFilter`, where you've had to `Box` the `set`'s closure with the key-to-predicate function, maybe both `DrainFilter`'s could share some common utility function. They would each pass in their respective predicates as a parameter, so the closure used in the `set` case doesn't have to be stored in any `struct`.

Reply:
Yes, they do share functions and a whole struct now.